src/trace/implementations/merge_batcher.rs

```diff
@@ -179,81 +143,235 @@ impl<D: Ord, T: Ord, R: Semigroup> MergeSorter<D, T, R> {
+        // the chains might be mis-ordered.
+        self.queue.sort_by_key(|chain| std::cmp::Reverse(chain.len()));
+
+        // Step 3: Merge chains that are within a power of two.
+        let mut index = self.queue.len().saturating_sub(1);
+        while index > 0 {
+            if self.queue[index-1].len() / 2 < self.queue[index].len() {
+                // Chains at `index-1` and `index` are within a factor of two, merge them.
+                let list1 = self.queue.remove(index-1);
+                let list2 = std::mem::take(&mut self.queue[index-1]);
+                self.queue[index-1] = self.merge_by(list1, list2);
+            }
+            index -= 1;
         }
-        self.queue_push(list);
+
+        self.account(self.queue.iter().flatten(), 1);
     }
 
-    #[inline(never)]
-    pub fn finish_into(&mut self, target: &mut Vec<Vec<(D, T, R)>>) {
-        while self.queue.len() > 1 {
-            let list1 = self.queue_pop().unwrap();
-            let list2 = self.queue_pop().unwrap();
-            let merged = self.merge_by(list1, list2);
-            self.queue_push(merged);
+    /// Extract all data that is not in advance of `upper`. Record the lower bound of the remaining
+    /// data's time in `frontier`.
+    fn extract_into(
+        &mut self,
+        upper: AntichainRef<T>,
+        frontier: &mut Antichain<T>,
+    ) -> Vec<Vec<(D, T, R)>> {
+        // Flush pending data
+        crate::consolidation::consolidate_updates(&mut self.pending);
+        self.flush_pending();
+
+        let mut keep = self.empty();
+        let mut keep_frontier = Antichain::new();
+        let mut ship = self.empty();
+        let mut ship_list = Vec::default();
+
+        self.account(self.queue.iter().flatten(), -1);
+
+        // Walk all chains, separate ready data from data to keep.
+        for mut chain in std::mem::take(&mut self.queue).drain(..) {
+            let mut block_list = Vec::default();
+            let mut keep_list = Vec::default();
+            for (block_frontier, mut block) in chain.drain(..) {
+                // Is any data ready to be shipped?
+                let any = !PartialOrder::less_equal(&upper, &block_frontier.borrow());
+                // Is all data ready to be shipped?
+                let all = any && block.iter().all(|(_, t, _)| !upper.less_equal(t));
+
+                if all {
+                    // All data is ready, push what we accumulated, stash whole block.
+                    if !ship.is_empty() {
+                        block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+                    }
+                    block_list.push((Antichain::new(), block));
+                } else if any {
+                    // Iterate block, sorting items into ship and keep
+                    for datum in block.drain(..) {
+                        if upper.less_equal(&datum.1) {
+                            frontier.insert_ref(&datum.1);
```

Comment:
This appears to be the only (?) place we update `frontier`, even though when we ship none of the updates (the `else` case below) we still want to reflect those times in the overall frontier.